src/math/matrix.c

```diff
@@ -3,93 +3,98 @@
+	/* Check for integer overflows during the shift operation. */
+	if (shift_minus_one < -1 || shift_minus_one > 62)
+		return -ERANGE;
 
-		return 0;
-	}
+	const int shift = shift_minus_one + 1;
+
+	/* Calculate offset for rounding. The offset is only needed when shift is
+	 * non-negative (> 0). Since shift = shift_minus_one + 1, the check for non-negative
+	 * value is (shift_minus_one >= -1)
+	 */
+	const int64_t offset = (shift_minus_one >= -1) ? (1LL << shift_minus_one) : 0;
 
 	for (i = 0; i < a->rows; i++) {
 		for (j = 0; j < b->columns; j++) {
-			s = 0;
+			acc = 0;
+			/* Position pointer x at the beginning of the ith row in matrix A */
 			x = a->data + a->columns * i;
+			/* Position pointer y at the beginning of the jth column in matrix B */
 			y = b->data + j;
 			for (k = 0; k < b->rows; k++) {
-				s += (int32_t)(*x) * (*y);
-				x++;
+				/* Multiply elements and add to sum */
+				acc += (int64_t)(*x++) * (*y);
+				/* Move pointer y to the next element in the column */
 				y += y_inc;
 			}
-			*z = (int16_t)(((s >> shift_minus_one) + 1) >> 1); /*Shift to Qx.y */
-			z++;
+			/* If shift == 0, then shift_minus_one == -1, which means the data is Q16.0
+			 * Otherwise, add the offset before the shift to round up if necessary
+			 */
+			*z++ = (shift == 0) ? (int16_t)acc : (int16_t)((acc + offset) >> shift);
 		}
 	}
+
 	return 0;
 }
 
 int mat_multiply_elementwise(struct mat_matrix_16b *a, struct mat_matrix_16b *b,
 			     struct mat_matrix_16b *c)
-{	int64_t p;
+{
+	/* check for NULL pointers */
+	if (!a || !b || !c)
+		return -EINVAL;
+	/* Validate dimensions match for elementwise multiplication */
+	if (a->columns != b->columns || a->rows != b->rows)
+		return -EINVAL;
+
+	const int total_elements = a->rows * a->columns;
+	/* Calculate shift for result Q format */
+	const int shift = a->fractions + b->fractions - c->fractions;
+	/* Check for integer overflows during the shift operation. */
+	if (shift < -1 || shift > 62)
+		return -ERANGE;
+	/* Offset for rounding */
+	const int64_t offset = (shift >= 0) ? (1LL << (shift - 1)) : 0;
+
 	int16_t *x = a->data;
 	int16_t *y = b->data;
 	int16_t *z = c->data;
+	int64_t acc;
 	int i;
-	const int shift_minus_one = a->fractions + b->fractions - c->fractions - 1;
 
-	if (a->columns != b->columns || b->columns != c->columns ||
-	    a->rows != b->rows || b->rows != c->rows) {
-		return -EINVAL;
-	}
-
-	/* If all data is Q0 */
-	if (shift_minus_one == -1) {
-		for (i = 0; i < a->rows * a->columns; i++) {
-			*z = *x * *y;
-			x++;
-			y++;
-			z++;
+	/* When no shifting is required (shift -1 indicated), simply multiply */
+	if (shift == -1) {
+		for (i = 0; i < total_elements; i++)
+			z[i] = x[i] * y[i]; /* Direct elementwise multiplication */
+	} else {
+		/* General case with shifting and offset for rounding */
+		for (i = 0; i < total_elements; i++) {
+			acc = (int64_t)x[i] * (int64_t)y[i]; /* Cast to int64_t to avoid overflow */
```

Comment:
I really wonder how 64 bit multiply can be faster than original 32 bits. This code avoids one shift but still, it's quite surprising. Did you test the non-Q0 case for cycles?

Reply:
You're right. Multiplication in 64 bits does not yield faster results than multiplication in 32 bits; instead, it requires twice as many resources (2x more multipliers).

My last check-in wasn't very successful. I've since fixed the error and included a summary of the results for the most recent check-in.

Thank you